### DISCUSSION TITLE
How does broadphase collision detection implement?

### DISCUSSION THREAD

**Discussion by acrlw** [dyn4j/dyn4j#196] [Q&A]
I am recently learning the implementation of broadphase collision filtering. I have seen the [source code](https://github.com/dyn4j/dyn4j/blob/master/src/main/java/org/dyn4j/collision/broadphase/DynamicAABBTree.java) and read the [publication](https://box2d.org/files/ErinCatto_DynamicBVH_Full.pdf) of Box2D. And also I have implement my [own version](https://github.com/AngryAccelerated/Physics2D/blob/master/source/collision/broadphase/dbvh.cpp) of dynamic tree. However, when I tested my tree and found that the time complexity of potential collision test is O(n * nlogn), which means O(n * n). 

These are steps what I use:

#### Step 1: Get Fat AABB
1. Get the thin AABB of body.
2. Scale it up with a specific factor (such as 1.2) to get fat AABB.
3. Insert it into Dynamic AABB tree.

#### Step 2: Insert
1. Get the AABB and make a new node
2. If there is no root node, make it root node and return.
3. If there is only root node, copy root node and create new node, calculate union set and assign to root node.
4. If there are more than two nodes, calculate the cost of constructing AABB of every leaf node and new node. Choose the minimum cost and merge new node with minimum cost node.

##### The calculation of cost
The same as [publication](https://box2d.org/files/ErinCatto_DynamicBVH_Full.pdf).

#### Step 3: Balance
1. Get the node to be balanced.
2. Calculate the height of node->leftChild and node->rightChild.

##### If leftHeight > rightHeight
1. Calculate height of node->leftChild->leftChild and node->leftChild->rightChild.
2 If left > right, then do left rotation.
3. If right > left, do right rotation first then left rotation.
##### If rightHeight > leftHeight
1. Calculate height of node->rightChild->leftChild and node->rightChild->rightChild.
2. If right > left, then do right rotation.
3. If left > right, do left rotation first then right rotation.

3. Balance left child and right child of node recursively untill node is leaf node.

#### Step 4: Update The AABB Of Branch
1. Get every leaf node.
2. Choose a leaff node.
3. Get the parent node.
4. Calculate the union set of left child node right child node, update the AABB of parent node.
5. Recursively Call function update() until node is root node.

But the real running result is:
1. 50 Bodies, Compare 442 times, Get 22 pairs.
2. 100 Bodies, Compare 777 times, Get 65 pairs.
3. 1000 Bodies, Compare 47715 times, Get 1352 pairs.
4. 1500 Bodies, Compare 641878 times, Get 12884 pairs.

All of them are rectangles. The width and height of rectangle are the same, 0.1, and rectangles are randomly distributed uniformly in 2d space, which is from (-9, 9) to (9, -9). Angle is from 360 degree to -360 degree.
The result looks like O(n * nlogn), but I have no idea how to optimize the construction of dynamic AABB tree.So does the step of balancing need to use Surface Area Heuristic to get cost then make decision? Should I optimize tree every time step?

**Comment by wnbittle** [comment]
A binary tree has O(log n) search time in non-degenerate cases.  We're searching a non-degenerate binary tree too, but we're performing one search per body in the scene.  This makes our complexity O(n log n) right out of the gate.

Now, the next thing to consider is that the search is not a _standard_ binary search.  A standard binary search would stop when it reaches a leaf node.  In our case, we can't stop, all the parents were overlapping and we need to check their subtrees for collisions.  Imagine the following tree where the leaves are the body AABBs and all the other nodes are their unioned AABBs up the tree:

```
           1
      /         \
     2           3
   /   \       /   \
  4     5     6     7
 / \   / \   / \   / \
A   B C   D E   F G   H
```

Which implies:
* 1 AABB contains all AABBs
* 2 AABB contains 4 and 5 AABBs
* 4 AABB contains A and B AABBs
* 5 AABB contains C and D AABBs
* and so on...

Therefore, if an AABB overlaps 1, 2, 4, and A, we still need to walk back up the tree and test B, 5, and 3.  If 5 or 3 overlap with the AABB, then I need to test their subtrees as well.  So the search complexity isn't strictly O(n log n).  For an AABB that's huge, you might need to iterate the whole tree.

Here's some numbers for comparison:

bodies | O(n^2) | O(n log n) | yours | dyn4j
----------|----------|----|----|----
50 | 2,500 | 282 | 442 | 866
100 | 10,000 | 664 | 777 | 2,330
1000 | 1,000,000 | 9,965 | 47,715 | 38,900
1500 | 2,250,000 | 15,826 | 641,878 | 66,736
6100 | 37,210,000 | 76,705 |  | 452,172

The numbers will depend on how you built your tree and the layout of the scene.  My numbers are based on N bodies randomly distributed in a 400 m grid sized randomly between 0.5 and 2.5m.  All bodies were static with exception of the 6,100 scenario where 100 were dynamic.

Now, also consider what we're trying to do here - we're NOT just trying to accelerate the broadphase collision detection.  We do want that, but we also want to accelerate raycasts.  So while a Sweep and Prune broadphase might have better broadphase complexity, it's has worse complexity wrt. raycasting.

Now, this is where things get crazy.  Remember the notion of expand AABBs?  It's NOT just to avoid updates to the broadphase acceleration structure.  [See here](https://dyn4j.org/2021/05/2021-05-02-a-faster-broadphase-in-dyn4j-4-0-0/) for more details on the optimization that's available and some results.  Obviously, I'm comparing against a best case O(n log n), but the concept is what matters.  For example, in the scenario above with 6,100 bodies _and_ this optimization implemented, it was on average around 5k comparisons instead of 400k...

**Reply by acrlw** [reply]
Oh my mistake. The number of bodies should be 5000:
5000 Bodies, Compare 641878 times, Get 12884 pairs.

And I also calculate another data:
Randomly distributed uniformly in (-400, 400) to (400, -400) :
1500 Bodies, Compare 81385 times, Get 6 pairs 

It looks much better.

Thanks for your reply! 😁
